File: packages/scheduling-engine/src/overflow/tier-types.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — Shared types for overflow tier functions
// ═══════════════════════════════════════════════════════════

import type { DecisionRegistry } from '../decisions/decision-registry.js';
import type { AdvanceAction, Block, MoveAction } from '../types/blocks.js';
import type { DecisionEntry } from '../types/decisions.js';
import type { EOp } from '../types/engine.js';
import type { FeasibilityReport } from '../types/infeasibility.js';

/** Mutable state threaded through tier functions */
export interface TierState {
  blocks: Block[];
  schedResult: {
    blocks: Block[];
    decisions: DecisionEntry[];
    registry: DecisionRegistry;
    feasibilityReport: FeasibilityReport;
  };
  autoMoves: MoveAction[];
  autoAdvances: AdvanceAction[];
}

/** Scheduling function signature used by all tiers */
export type RunScheduleFn = (
  moves: MoveAction[],
  advances?: AdvanceAction[],
) => {
  blocks: Block[];
  decisions: DecisionEntry[];
  registry: DecisionRegistry;
  feasibilityReport: FeasibilityReport;
};

/** Shared context passed to tier functions */
export interface TierContext {
  ops: EOp[];
  userMoves: MoveAction[];
  mSt: Record<string, string>;
  workdays: boolean[];
  twinPartnerMap: Map<string, string>;
  thirdShift?: boolean;
  machineTimelines?: Record<string, import('../types/failure.js').ResourceTimeline>;
  runSchedule: RunScheduleFn;
  /** Schedule with load leveling enabled (slower, redistributes overflow). Used by Tier 3 bulk trials. */
  runScheduleWithLeveling?: RunScheduleFn;
}

File: apps/backend/src/domain/scheduling/overflow/tier_types.py
"""Shared types for overflow tier functions — port of overflow/tier-types.ts."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field

from ..types import AdvanceAction, Block, EOp, MoveAction, ScheduleResult

# RunScheduleFn: (moves, advances?) -> ScheduleResult
RunScheduleFn = Callable[..., ScheduleResult]


@dataclass
class TierState:
    """Mutable state threaded through tier functions."""

    blocks: list[Block]
    sched_result: ScheduleResult
    auto_moves: list[MoveAction] = field(default_factory=list)
    auto_advances: list[AdvanceAction] = field(default_factory=list)


@dataclass
class TierContext:
    """Shared context passed to tier functions."""

    ops: list[EOp]
    user_moves: list[MoveAction]
    m_st: dict[str, str]
    workdays: list[bool]
    twin_partner_map: dict[str, str]
    third_shift: bool = False
    run_schedule: RunScheduleFn = None  # type: ignore[assignment]
    run_schedule_with_leveling: RunScheduleFn | None = None
